# AS004: SKILL.md Body Length Limit

## Source

**Official Documentation**: [Progressive Disclosure Patterns - Agent Skills Best Practices](https://platform.claude.com/docs/en/agents-and-tools/agent-skills/best-practices#progressive-disclosure-patterns)

## Rule Description

Keep SKILL.md body under 500 lines for optimal performance. Split content into separate files when approaching this limit.

## Requirements

The markdown body content in SKILL.md (excluding frontmatter) must be under 500 lines.

## Rationale

SKILL.md serves as an overview that points Claude to detailed materials as needed, like a table of contents in an onboarding guide. When SKILL.md exceeds 500 lines:

1. **Context window pressure**: Large SKILL.md files consume significant context that could be used for conversation history and reasoning
2. **Discovery inefficiency**: Claude must load the entire file when the skill is triggered, even if only a small portion is relevant
3. **Progressive disclosure breaks down**: The point of progressive disclosure is to load details only when needed

By keeping SKILL.md concise and splitting detailed content into referenced files, Claude can:
- Load the overview quickly
- Read specific reference files only when needed
- Maintain more conversation context

## Good Example Structure

```markdown
---
name: pdf-processing
description: Extracts text and tables from PDF files...
---

# PDF Processing

## Quick start
[Basic examples here - 50-100 lines]

## Advanced features
**Form filling**: See [FORMS.md](FORMS.md)
**API reference**: See [REFERENCE.md](REFERENCE.md)
**Examples**: See [EXAMPLES.md](EXAMPLES.md)
```

Total SKILL.md: ~150 lines (well under 500)
Details split into FORMS.md, REFERENCE.md, EXAMPLES.md

## Bad Example

```markdown
---
name: pdf-processing
description: Extracts text...
---

# PDF Processing

[800 lines of comprehensive documentation, examples, API reference,
form-filling guides, troubleshooting, etc. all in one file]
```

## Detection

This rule:
1. Reads the SKILL.md file
2. Extracts content after frontmatter (between `---` delimiters)
3. Counts lines in the body content
4. Flags if body exceeds 500 lines

## Severity

- **Default**: Warning
- **Strict mode**: Error

## Remediation

When approaching or exceeding 500 lines:

1. **Identify separable content**: API references, advanced examples, troubleshooting guides
2. **Create reference files**: Move content to separate .md files
3. **Link from SKILL.md**: Add clear links to the reference files
4. **Use progressive disclosure patterns**: Organize by domain, feature, or complexity level

## Related Rules

- AS006: References should be one level deep
- AS008: Table of contents required for files > 100 lines
