# AS008: Table of Contents for Long Files

## Source

**Official Documentation**: [Structure Longer Reference Files - Agent Skills Best Practices](https://platform.claude.com/docs/en/agents-and-tools/agent-skills/best-practices#structure-longer-reference-files-with-table-of-contents)

## Rule Description

For reference files longer than 100 lines, include a table of contents at the top. This ensures Claude can see the full scope of available information even when previewing with partial reads.

## Requirements

Files exceeding 100 lines should include a table of contents (TOC) near the beginning (within the first 20 lines of content).

## Rationale

**Progressive loading behavior**: When Claude encounters a referenced file, it may initially preview it with commands like `head -100` rather than reading the entire file. This is efficient for small files but problematic for large reference documents.

**Without a TOC**:
- Claude sees only the first 100 lines
- May miss sections it needs
- Wastes time searching or reading unnecessary content

**With a TOC**:
- Claude immediately sees all available sections
- Can directly navigate to relevant sections
- Makes informed decisions about what to read

**Performance impact**:
- **Without TOC**: "Let me read the whole file to see what's there"
- **With TOC**: "I need section 5, let me jump there directly"

## Good Example

````markdown
# API Reference

## Table of Contents

1. [Authentication](#authentication)
2. [Core Methods](#core-methods)
   - [Create Document](#create-document)
   - [Update Document](#update-document)
   - [Delete Document](#delete-document)
3. [Advanced Features](#advanced-features)
4. [Error Handling](#error-handling)
5. [Rate Limits](#rate-limits)

## Authentication

[Content continues for 150+ lines]
````

Even with a 150-line file, the TOC (in first 10 lines) shows Claude all available sections.

## Bad Example

````markdown
# API Reference

## Authentication

[50 lines about authentication]

## Core Methods

[50 lines about core methods]

## Advanced Features

[60 lines about advanced features]
````

Without a TOC, Claude previewing with `head -100` would miss "Advanced Features" entirely.

## Detection

This rule:
1. Identifies markdown files > 100 lines in the skill directory
2. Checks for TOC indicators in the first 20 lines:
   - Headers containing "table of contents", "contents", or "toc" (case-insensitive)
   - Multiple markdown links to anchors (pattern: `[text](#anchor)`)
   - Counts `](#` occurrences > 3 as implicit TOC

## Severity

- **Default**: Warning
- **Strict mode**: Error

## TOC Format Options

**Explicit header**:
```markdown
## Table of Contents
- [Section 1](#section-1)
- [Section 2](#section-2)
```

**Implicit** (multiple anchor links):
```markdown
This guide covers [Authentication](#auth), [Methods](#methods), and [Examples](#examples).
```

**Nested**:
```markdown
## Contents
1. [Overview](#overview)
2. [API Reference](#api-reference)
   - [GET endpoints](#get-endpoints)
   - [POST endpoints](#post-endpoints)
```

## Remediation

For files > 100 lines without a TOC, add one near the beginning:

```markdown
---
name: pdf-processing
description: ...
---

# PDF Processing

## Table of Contents

1. [Quick Start](#quick-start)
2. [Form Filling](#form-filling)
3. [Table Extraction](#table-extraction)
4. [OCR Configuration](#ocr-configuration)
5. [Troubleshooting](#troubleshooting)

## Quick Start

[Content continues...]
```

## Related Rules

- AS004: SKILL.md body < 500 lines (prevents files from getting too long)
- AS006: References one level deep (ensures TOC files are directly accessible)
- AS020: TOC completeness (ensures TOC matches actual headers)
