# AS001: Name Format Validation

## Source

**Official Documentation**: [Naming Conventions - Agent Skills Best Practices](https://platform.claude.com/docs/en/agents-and-tools/agent-skills/best-practices#naming-conventions)

## Rule Description

The skill `name` field must conform to strict formatting requirements to ensure discoverability and consistent behavior across the Claude platform.

## Requirements

1. **Maximum Length**: 64 characters
2. **Character Set**: Only lowercase letters, numbers, and hyphens
3. **No XML Tags**: Cannot contain angle brackets (`<` or `>`)
4. **No Reserved Words**: Cannot contain "anthropic" or "claude"
5. **Hyphen Rules**:
   - Cannot start or end with a hyphen
   - Cannot contain consecutive hyphens (`--`)

## Rationale

- **Lowercase requirement**: Ensures consistent naming across filesystems and platforms
- **Limited character set**: Prevents encoding issues and ensures URL-safe names
- **No XML tags**: Prevents injection attacks and parsing errors
- **Reserved words**: Protects official Anthropic/Claude branding and prevents confusion
- **Hyphen rules**: Ensures valid identifier format and prevents parsing ambiguities

## Good Examples

```yaml
name: processing-pdfs
```

```yaml
name: analyzing-spreadsheets
```

```yaml
name: managing-databases
```

```yaml
name: testing-code
```

## Bad Examples

```yaml
# Too long (exceeds 64 characters)
name: this-is-a-really-long-skill-name-that-exceeds-the-maximum-allowed-length-of-64-characters

# Contains uppercase
name: ProcessingPDFs

# Contains underscore
name: processing_pdfs

# Contains reserved word
name: claude-helper

# Contains XML tag
name: <processing>pdfs

# Starts with hyphen
name: -processing-pdfs

# Contains consecutive hyphens
name: processing--pdfs
```

## Detection

This rule checks the skill metadata `name` field against all format requirements.

## Severity

- **Default**: Warning
- **Strict mode**: Error

## Related Rules

- AS015: Prefer gerund naming (verb-ing pattern)
- AS016: Avoid reserved words (enforces same reserved word list)
