# AS007: Descriptive File Naming

## Source

**Official Documentation**: [Technical Architecture Notes - Agent Skills Best Practices](https://platform.claude.com/docs/en/agents-and-tools/agent-skills/best-practices#technical-architecture-notes)

## Rule Description

Name files descriptively to indicate their content. Avoid generic names like `doc1.md`, `doc2.md`, `file.md`, `script.py`, or `helper.py`.

## Requirements

File names should clearly indicate their purpose or content:
- ✓ **Good**: `form_validation_rules.md`, `pdf_extraction.py`, `api_reference.md`
- ✗ **Avoid**: `doc1.md`, `doc2.md`, `file2.md`, `script.py`, `helper.py`, `utils.md`

## Rationale

**Claude navigates your skill directory like a filesystem**. Descriptive file names help Claude:

1. **Understand content without reading**: `billing_queries.md` is clearly about billing
2. **Choose the right file**: When looking for form validation, `form_validation_rules.md` > `doc3.md`
3. **Build mental model**: `reference/finance.md`, `reference/sales.md` shows organization
4. **Avoid unnecessary reads**: Clear names prevent loading wrong files

**Why generic names hurt**:
- `doc1.md`, `doc2.md`, `doc3.md` → Which one has the information? Claude must read all three
- `helper.py`, `utils.py` → What do they help with? No indication of purpose
- `file.md` → Tells you nothing about content

Generic names waste context tokens and slow down task completion.

## Good Examples

```
pdf-skill/
├── SKILL.md
├── form_validation_rules.md
├── ocr_configuration.md
├── table_extraction_guide.md
└── scripts/
    ├── analyze_form_fields.py
    ├── extract_tables.py
    └── validate_signatures.py
```

Each file name clearly indicates its content and purpose.

## Bad Examples

```
pdf-skill/
├── SKILL.md
├── doc1.md          # What's in doc1?
├── doc2.md          # What's in doc2?
├── guide.md         # Guide to what?
└── scripts/
    ├── script1.py   # What does this do?
    ├── script2.py   # What about this?
    └── helper.py    # Helps with what?
```

Claude must read each file to understand what it contains.

## Detection

This rule scans all files in the skill directory for generic naming patterns:

**Flagged patterns**:
- `doc[0-9].md`, `doc[0-9][0-9].md` (doc1.md, doc2.md, etc.)
- `file[0-9].md`, `file[0-9][0-9].md`
- `script[0-9].py`, `script[0-9][0-9].py`
- `helper.py`, `helper.js`, `helper.sh`
- `utils.md`, `utils.py`, `utils.js`
- `misc.md`, `other.md`, `temp.md`

**Exceptions**:
- `SKILL.md`, `README.md`, `LICENSE.md` (standard files)
- Files in `.git/`, `node_modules/`, `__pycache__/` (ignored directories)

## Severity

- **Default**: Warning
- **Strict mode**: Error

## Remediation

Rename files to describe their content:

```diff
- doc1.md → form_validation_rules.md
- doc2.md → api_reference.md
- script1.py → extract_tables.py
- helper.py → pdf_form_analyzer.py
```

Update references in SKILL.md:

```diff
- See [documentation](doc1.md)
+ See [form validation rules](form_validation_rules.md)
```

## Naming Guidelines

**For reference files**:
- Describe the topic: `api_reference.md`, `configuration_guide.md`
- Include domain if organizing by domain: `finance_queries.md`, `sales_metrics.md`

**For scripts**:
- Use verb + noun: `extract_tables.py`, `validate_form.py`, `analyze_pdf.py`
- Describe the operation: `merge_documents.sh`, `convert_to_text.py`

**For examples**:
- Indicate use case: `example_invoice_processing.md`, `basic_usage.md`

## Related Rules

- AS005: Use forward slashes in paths
- AS006: References one level deep
- AS008: TOC for files > 100 lines
