# AS003: Third-Person Voice

## Source

**Official Documentation**: [Writing Effective Descriptions - Agent Skills Best Practices](https://platform.claude.com/docs/en/agents-and-tools/agent-skills/best-practices#writing-effective-descriptions)

## Rule Description

Always write skill descriptions in third person. The description is injected into the system prompt, and inconsistent point-of-view can cause discovery problems.

## Requirements

Avoid first and second person pronouns:
- **Avoid**: "I", "we", "our", "my"
- **Avoid**: "you", "your"

## Rationale

Skill descriptions are injected directly into Claude's system prompt. Using first or second person creates point-of-view inconsistencies that can confuse the skill selection process. Third-person descriptions maintain a consistent, objective tone that clearly describes capabilities without introducing ambiguous pronouns.

## Good Examples

```yaml
description: Processes Excel files and generates reports
```

```yaml
description: Extracts text from PDF documents and analyzes content
```

```yaml
description: Manages database migrations and schema updates
```

## Bad Examples

```yaml
# First person ("I")
description: I can help you process Excel files
```

```yaml
# Second person ("you")
description: You can use this to process Excel files
```

```yaml
# First person plural ("we")
description: We extract text from PDF documents
```

```yaml
# Possessive second person ("your")
description: Helps analyze your spreadsheet data
```

## Detection

This rule scans the description field for common first and second person pronouns:
- "i " (case-insensitive, with trailing space)
- "you "
- "we "
- "our "
- "my "
- "your "

## Severity

- **Default**: Warning
- **Strict mode**: Error

## Related Rules

- AS002: Description validation (ensures description exists and is well-formed)
- AS014: Description should include usage triggers
