# AS005: Forward Slashes in Paths

## Source

**Official Documentation**: [Avoid Windows-style Paths - Agent Skills Best Practices](https://platform.claude.com/docs/en/agents-and-tools/agent-skills/best-practices#avoid-windows-style-paths)

## Rule Description

Always use forward slashes (`/`) in file paths, even on Windows. Never use backslashes (`\`).

## Requirements

All file path references in skill content must use forward slashes:
- ✓ **Good**: `scripts/helper.py`, `reference/guide.md`
- ✗ **Avoid**: `scripts\helper.py`, `reference\guide.md`

## Rationale

1. **Cross-platform compatibility**: Forward slashes work on all platforms (Windows, macOS, Linux)
2. **Consistency**: Skills should work identically regardless of where they're used
3. **Escape character conflicts**: Backslashes are escape characters in many contexts, causing parsing issues
4. **Claude's filesystem navigation**: Claude navigates skill directories like a Unix filesystem

Even on Windows, modern path APIs accept forward slashes. Using backslashes:
- Breaks skills when shared across platforms
- Creates markdown escaping issues (`\n`, `\t`, etc.)
- Confuses Claude's file navigation

## Good Examples

```markdown
See [FORMS.md](FORMS.md) for form filling
See [reference/api.md](reference/api.md) for API docs
Run `python scripts/analyze.py`
```

```python
# Code example with forward slashes
with open("data/input.csv", "r") as f:
    data = f.read()
```

## Bad Examples

```markdown
# Backslashes (Windows-style)
See [FORMS.md](FORMS.md) for form filling  # OK - no path
See [reference\api.md](reference\api.md)    # BAD - backslash

Run `python scripts\analyze.py`             # BAD - backslash
```

```python
# Code example with backslashes
with open("data\input.csv", "r") as f:     # BAD
    data = f.read()
```

## Detection

This rule scans all skill content (SKILL.md and referenced files) for backslash characters (`\`) that appear to be path separators. It looks for patterns like:
- `word\word`
- `path\to\file`

Excludes legitimate uses:
- Escape sequences in code (`\n`, `\t`, `\"`)
- Regex patterns (`\d`, `\w`)
- LaTeX/math notation

## Severity

- **Default**: Warning
- **Strict mode**: Error

## Remediation

Replace all backslashes in paths with forward slashes:

```diff
- See [guide](reference\guide.md)
+ See [guide](reference/guide.md)

- Run `python scripts\helper.py`
+ Run `python scripts/helper.py`
```

## Platform Note

On Windows, both Python and Node.js (and most modern tools) accept forward slashes in paths, so this change is safe and improves portability.

## Related Rules

- AS006: References should be one level deep
- AS007: Use descriptive file naming
