# AS006: Reference Depth Limit

## Source

**Official Documentation**: [Avoid Deeply Nested References - Agent Skills Best Practices](https://platform.claude.com/docs/en/agents-and-tools/agent-skills/best-practices#avoid-deeply-nested-references)

## Rule Description

Keep references one level deep from SKILL.md. All reference files should link directly from SKILL.md to ensure Claude reads complete files when needed.

## Requirements

**Maximum reference depth**: 1 level
- ✓ SKILL.md → REFERENCE.md (allowed)
- ✗ SKILL.md → GUIDE.md → DETAILS.md (too deep)

## Rationale

Claude may partially read files when they're referenced from other referenced files. When encountering nested references, Claude might use commands like `head -100` to preview content rather than reading entire files, resulting in incomplete information.

**Why this happens**:
1. SKILL.md is fully loaded when the skill is triggered
2. Files directly referenced from SKILL.md are fully loaded when accessed
3. Files referenced from *those* files may only be previewed (not fully loaded)
4. Result: Nested reference content may be incomplete

**Impact of deep nesting**:
- Critical information may be truncated
- Users receive incomplete guidance
- Skill behavior becomes unpredictable

## Good Example: One Level Deep

```markdown
# SKILL.md
---
name: pdf-processing
description: PDF processing tools
---

## Features

**Basic usage**: Extract text with pdfplumber
**Advanced forms**: See [FORMS.md](FORMS.md)
**API reference**: See [REFERENCE.md](REFERENCE.md)
**Troubleshooting**: See [TROUBLESHOOTING.md](TROUBLESHOOTING.md)
```

All references link directly from SKILL.md. When Claude needs form handling details, it fully loads FORMS.md.

## Bad Example: Too Deep

```markdown
# SKILL.md
---
name: pdf-processing
description: PDF processing tools
---

See [GUIDE.md](GUIDE.md) for all features
```

```markdown
# GUIDE.md
For advanced forms, see [FORMS.md](FORMS.md)
For API reference, see [REFERENCE.md](REFERENCE.md)
```

```markdown
# FORMS.md
Here's the actual information...
```

Problem: FORMS.md is **two levels deep** (SKILL.md → GUIDE.md → FORMS.md). Claude may only preview it with `head -100`, missing critical content.

## Detection

This rule scans SKILL.md for file references and checks those referenced files for additional file references. It flags any references found in already-referenced files.

Detection pattern:
1. Extract all markdown links from SKILL.md: `[text](file.md)`
2. For each linked file, scan for additional markdown links
3. Report any links found (indicating nested references)

## Severity

- **Default**: Warning
- **Strict mode**: Error

## Remediation

**Option 1: Flatten references** (recommended)

Move all important reference links directly into SKILL.md:

```markdown
# SKILL.md

**Basic usage**: [instructions here]
**Form filling**: See [FORMS.md](FORMS.md)
**API reference**: See [REFERENCE.md](REFERENCE.md)
**Examples**: See [EXAMPLES.md](EXAMPLES.md)
```

**Option 2: Inline intermediate content**

If GUIDE.md is small, inline it into SKILL.md:

```markdown
# SKILL.md

## Features

[Content from GUIDE.md here]

**Advanced topics**:
- Form filling: See [FORMS.md](FORMS.md)
- API reference: See [REFERENCE.md](REFERENCE.md)
```

## Related Rules

- AS004: SKILL.md body < 500 lines (ensures SKILL.md stays manageable)
- AS008: TOC required for files > 100 lines (helps navigation in reference files)
