# AS002: Description Validation

## Source

**Official Documentation**: [Writing Effective Descriptions - Agent Skills Best Practices](https://platform.claude.com/docs/en/agents-and-tools/agent-skills/best-practices#writing-effective-descriptions)

## Rule Description

The skill `description` field must be concise, informative, and properly formatted to enable effective skill discovery.

## Requirements

1. **Non-empty**: Description cannot be blank
2. **Maximum Length**: 1024 characters
3. **No XML Tags**: Cannot contain angle brackets (`<` or `>`)
4. **Should Include**:
   - What the skill does
   - When to use it (triggers/contexts)

## Rationale

The description is critical for skill selection. Claude uses it to choose the right skill from potentially 100+ available skills. The description must provide enough detail for Claude to know when to select this skill, while the rest of SKILL.md provides implementation details.

**Why no XML tags**: Prevents injection attacks and ensures clean rendering in all contexts.

**Why 1024 character limit**: Balances detail with conciseness. Longer descriptions hurt discovery by cluttering the skill selection context.

## Good Examples

```yaml
description: Extract text and tables from PDF files, fill forms, merge documents. Use when working with PDF files or when the user mentions PDFs, forms, or document extraction.
```

```yaml
description: Analyze Excel spreadsheets, create pivot tables, generate charts. Use when analyzing Excel files, spreadsheets, tabular data, or .xlsx files.
```

```yaml
description: Generate descriptive commit messages by analyzing git diffs. Use when the user asks for help writing commit messages or reviewing staged changes.
```

## Bad Examples

```yaml
# Too vague
description: Helps with documents
```

```yaml
# Contains XML tags
description: Process <PDF> files and extract text
```

```yaml
# Empty description
description:
```

```yaml
# Too long (exceeds 1024 characters)
description: This skill is designed to help you process PDF files...
  [imagine this continues for many more sentences]
```

## Detection

This rule checks:
1. Description is not empty
2. Length does not exceed 1024 characters
3. No XML-like tags are present (regex match for `<[a-zA-Z][a-zA-Z0-9]*>`)

## Severity

- **Default**: Warning
- **Strict mode**: Error

## Related Rules

- AS003: Use third-person voice (applies to description content)
- AS014: Description should include usage triggers
