# AS009: MCP Tool References

## Source

**Official Documentation**: [MCP Tool References - Agent Skills Best Practices](https://platform.claude.com/docs/en/agents-and-tools/agent-skills/best-practices#mcp-tool-references)

## Rule Description

If your skill uses MCP (Model Context Protocol) tools, always use fully qualified tool names to avoid "tool not found" errors.

## Requirements

**Format**: `ServerName:tool_name`

Where:
- `ServerName` is the MCP server name
- `tool_name` is the tool name within that server

## Rationale

Without the server prefix, Claude may fail to locate the tool, especially when multiple MCP servers are available.

**Why this matters**:
1. **Namespace collision**: Multiple servers might have tools with similar names
2. **Tool discovery**: Claude needs to know which server provides the tool
3. **Error prevention**: Unqualified names lead to "tool not found" errors
4. **Explicit is better**: Clear dependencies make skills more maintainable

**Real-world scenario**:
- You have two MCP servers: `BigQuery` and `Snowflake`
- Both have a tool named `execute_query`
- Without prefix: `execute_query` → ambiguous, might use wrong server
- With prefix: `BigQuery:execute_query` → explicit, no ambiguity

## Good Examples

```markdown
Use `BigQuery:bigquery_schema` to get table schema
Use `GitHub:create_issue` to create issues
Use `Slack:send_message` to send notifications
```

```python
# Query BigQuery
result = use_tool("BigQuery:execute_query", {
    "query": "SELECT * FROM dataset.table"
})

# Create GitHub issue
issue = use_tool("GitHub:create_issue", {
    "title": "Bug report",
    "body": "Description"
})
```

## Bad Examples

```markdown
# Missing server prefix
Use `bigquery_schema` to get table schema     # BAD - which server?
Use `create_issue` to create issues           # BAD - ambiguous
Use `send_message` to send notifications      # BAD - could be many services
```

```python
# Unqualified tool names
result = use_tool("execute_query", {...})    # BAD - which database?
issue = use_tool("create_issue", {...})      # BAD - which issue tracker?
```

## Detection

This rule scans skill content for MCP tool references and checks for proper formatting:

**Patterns searched**:
- Inline code: `` `tool_name` ``
- Code blocks with tool invocations
- Terms like "use tool", "call", "invoke" followed by tool names

**Validation**:
- Tool names should contain `:` (colon separator)
- Format should be `Word:word_name` or `Word:wordName`
- Missing colon suggests unqualified tool name

**Heuristics**:
- Common MCP tool verbs: `get`, `create`, `update`, `delete`, `list`, `search`, `execute`
- Common server prefixes: `BigQuery`, `GitHub`, `Slack`, `Jira`, `Snowflake`, `Postgres`

## Severity

- **Default**: Warning
- **Strict mode**: Error

## Remediation

Add server prefix to all tool references:

```diff
- Use `bigquery_schema` to get table schema
+ Use `BigQuery:bigquery_schema` to get table schema

- Call `create_issue` to file bugs
+ Call `GitHub:create_issue` to file bugs

- result = use_tool("send_message", {...})
+ result = use_tool("Slack:send_message", {...})
```

## Finding Server Names

If unsure of the server name, check:
1. MCP server configuration (typically in `mcp.json` or similar)
2. MCP server documentation
3. Use `list_tools()` or equivalent to see available tools with server prefixes

## Related Rules

- AS013: Document required packages (should list required MCP servers)
