# AS015: Prefer Gerund Naming

## Source

**Official Documentation**: [Naming Conventions - Agent Skills Best Practices](https://platform.claude.com/docs/en/agents-and-tools/agent-skills/best-practices#naming-conventions)

## Rule Description

We recommend using **gerund form** (verb + -ing) for skill names, as this clearly describes the activity or capability the skill provides.

## Requirements

**Preferred pattern**: `verb-ing-object`
- processing-pdfs
- analyzing-spreadsheets
- managing-databases
- testing-code
- writing-documentation

**Acceptable alternatives**:
- Noun phrases: `pdf-processing`, `spreadsheet-analysis`
- Action-oriented: `process-pdfs`, `analyze-spreadsheets`

## Rationale

**Gerund names describe capabilities clearly**. The -ing form:

1. **Activity-focused**: Emphasizes what the skill DOES
2. **Consistent pattern**: Easy to scan and understand
3. **Natural language**: "processing PDFs" reads well
4. **Scales well**: Works for simple and complex skills

**Example collection**:
- `processing-pdfs` - immediately clear
- `analyzing-spreadsheets` - describes the action
- `managing-databases` - shows capability

vs.

- `pdf` - what about PDFs?
- `spreadsheet` - tool or action?
- `database` - too vague

## Good Examples

**Gerund form (recommended)**:

```yaml
name: processing-pdfs
description: Extract text and tables from PDF files
```

```yaml
name: analyzing-spreadsheets
description: Analyze Excel files and generate reports
```

```yaml
name: managing-databases
description: Database migration and schema management
```

```yaml
name: testing-code
description: Run unit tests and generate coverage reports
```

```yaml
name: writing-documentation
description: Generate API documentation from code
```

**Acceptable alternatives**:

```yaml
name: pdf-processing
description: Extract text and tables from PDF files
```

```yaml
name: spreadsheet-analysis
description: Analyze Excel files and generate reports
```

```yaml
name: database-management
description: Database migration and schema management
```

## Bad Examples

**Too vague**:

```yaml
name: helper
```

```yaml
name: utils
```

```yaml
name: tools
```

**Overly generic**:

```yaml
name: documents
```

```yaml
name: data
```

```yaml
name: files
```

**Non-descriptive**:

```yaml
name: skill-one
```

```yaml
name: my-skill
```

## Detection

This rule checks if skill names follow recommended patterns:

**Checks for gerund form**:
- Pattern: `*-ing-*` or `*ing-*`
- Examples: `processing-pdfs`, `analyzing-data`

**Checks for acceptable alternatives**:
- Noun-action pattern: `pdf-processing`, `data-analysis`
- Action-noun pattern: `process-pdfs`, `analyze-data`

**Flags generic names**:
- `helper`, `utils`, `tools` (too vague)
- `documents`, `data`, `files` (too generic)
- `skill-*`, `my-*` (non-descriptive)

## Severity

- **Default**: Warning (this is a style recommendation)
- **Strict mode**: Error

## Naming Guidelines

**Choose action verbs**:
- Processing, analyzing, managing, testing, writing
- Generating, validating, transforming, monitoring
- Creating, updating, deleting, querying

**Add object/context**:
- processing-**pdfs**
- analyzing-**spreadsheets**
- managing-**databases**

**Keep it concise**:
- ✓ `analyzing-sales-data`
- ✗ `analyzing-and-processing-sales-data-from-multiple-sources`

**Use hyphens for compound words**:
- ✓ `processing-pdf-forms`
- ✗ `processingpdfforms`

## Consistency in Collections

If you have multiple skills, use consistent patterns:

**Good - All gerund**:
- `processing-pdfs`
- `analyzing-spreadsheets`
- `managing-databases`

**Mixed (less ideal but acceptable)**:
- `pdf-processing` (noun-action)
- `analyze-spreadsheets` (verb-object)
- `managing-databases` (gerund)

**Bad - Inconsistent styles**:
- `processing-pdfs` (gerund)
- `spreadsheet` (just noun)
- `db-helper` (generic)

## Related Rules

- AS001: Name format validation (enforces technical requirements)
- AS016: Avoid reserved words (protects specific terms)
