# AS018: No Undocumented Magic Constants

## Source

**Official Documentation**: [Code and Scripts Checklist - Agent Skills Best Practices](https://platform.claude.com/docs/en/agents-and-tools/agent-skills/best-practices#evaluation-and-iteration)

## Rule Description

Avoid "voodoo constants" (Ousterhout's law). All configuration values and numeric constants must be documented with justification.

## Requirements

For all constants in code:
1. Explain WHY the value is what it is
2. Document the reasoning or source
3. Make constants self-documenting through naming and comments

## Rationale

**If you don't know the right value, how will Claude?** Undocumented constants create:

1. **Uncertainty**: Is this value correct or arbitrary?
2. **Fragility**: Fear of changing values that might be calibrated
3. **Debugging difficulty**: Hard to know if constants are the problem
4. **Knowledge loss**: Original reasoning disappears

**With documented constants**:
- Clear reasoning preserved
- Safe to adjust when needed
- Easier debugging and optimization
- Knowledge transfer

## Good Examples

**Self-documenting Python**:

```python
# HTTP requests typically complete within 30 seconds
# Extending to 60s accounts for slow networks and large responses
HTTP_TIMEOUT_SECONDS = 60

# PDF text extraction batch size
# Processing 10 pages at a time balances memory usage with performance
# Tested with 1000-page documents: 10 pages = optimal throughput
BATCH_SIZE_PAGES = 10

# Retry configuration
# 3 attempts allows for transient network issues without excessive delay
# Exponential backoff: 1s, 2s, 4s = 7s total maximum delay
MAX_RETRIES = 3
INITIAL_RETRY_DELAY_SECONDS = 1
```

**Self-documenting JavaScript**:

```javascript
// API rate limit: 100 requests per minute per their documentation
// https://api.example.com/docs/rate-limits
const MAX_REQUESTS_PER_MINUTE = 100;

// Minimum delay between requests to stay under rate limit
// 60000ms / 100 requests = 600ms minimum delay
const MIN_REQUEST_DELAY_MS = 600;

// File size limit: API rejects files > 10MB
// Source: https://api.example.com/docs/upload-limits
const MAX_FILE_SIZE_BYTES = 10 * 1024 * 1024;
```

**Configuration with comments**:

```python
config = {
    # Database connection pool
    # 20 connections handles peak load (measured at 15 concurrent queries)
    'pool_size': 20,

    # Query timeout
    # 30s allows complex analytical queries while preventing runaway queries
    # 95th percentile query time: 8 seconds (measured)
    'query_timeout_seconds': 30,

    # Cache TTL
    # 5 minutes balances freshness with API cost
    # Data updates every 10 minutes, so 5m cache is safe
    'cache_ttl_seconds': 300,
}
```

## Bad Examples

**Undocumented magic numbers**:

```python
timeout = 42  # Why 42? Is this tested? Can it be changed?

batch_size = 17  # Why 17? Seems arbitrary

max_retries = 5  # Why 5? Too many? Too few?
```

**Unexplained configuration**:

```python
config = {
    'size': 1024,  # Why 1024?
    'delay': 500,  # Why 500ms?
    'limit': 73,   # Why 73? Very specific...
}
```

**No justification**:

```javascript
const THRESHOLD = 0.85;  // What threshold? For what?
const BUFFER = 2048;     // Why this size?
const FACTOR = 1.5;      // What is this factor for?
```

## Detection

This rule scans code for numeric constants and checks for documentation:

**Looks for**:
- Assignment of numbers: `CONSTANT = 42`
- Configuration values: `timeout: 30`
- Unexplained thresholds: `if value > 100`

**Checks for documentation**:
- Comment on same or preceding line
- Variable name indicates purpose
- Nearby documentation explaining value

**Flags**:
- Unexplained numeric literals
- Constants without comments
- Configuration values without rationale

## Severity

- **Default**: Warning
- **Strict mode**: Error

## What Counts as Documentation

**Sufficient**:

```python
# HTTP timeout: 30s per API documentation
TIMEOUT = 30
```

```python
# Batch size: tested optimal value for memory/performance balance
BATCH_SIZE = 100
```

**Insufficient**:

```python
# Timeout value
TIMEOUT = 30  # Describes what, not why
```

```python
BATCH_SIZE = 100  # No comment at all
```

## Common Constants That Need Documentation

**Timeouts**:
```python
# HTTP requests complete in 30s (99th percentile from monitoring)
HTTP_TIMEOUT_SECONDS = 30
```

**Buffer/batch sizes**:
```python
# Process 1000 records per batch (optimal for 8GB memory constraint)
BATCH_SIZE = 1000
```

**Rate limits**:
```python
# API allows 60 requests/minute per documentation
MAX_REQUESTS_PER_MINUTE = 60
```

**Thresholds**:
```python
# Confidence threshold: 0.85 based on precision/recall testing
# Precision: 0.92, Recall: 0.88 at this threshold
CONFIDENCE_THRESHOLD = 0.85
```

**Retry/backoff**:
```python
# Retry 3 times for transient errors (network, rate limit)
# Backoff: 1s, 2s, 4s exponential
MAX_RETRIES = 3
```

## Acceptable Constants Without Documentation

**Named constants** (self-explanatory):
```python
SECONDS_PER_MINUTE = 60
BYTES_PER_MEGABYTE = 1024 * 1024
HTTP_STATUS_OK = 200
```

**Standard values**:
```python
DEFAULT_PORT = 80
HTTPS_PORT = 443
```

**Industry standards**:
```python
TCP_KEEP_ALIVE = 7200  # RFC 1122 standard
```

## Related Rules

- AS017: Scripts have error handling (error thresholds should be documented)
