# AS013: Document Required Packages

## Source

**Official Documentation**: [Code and Scripts Checklist - Agent Skills Best Practices](https://platform.claude.com/docs/en/agents-and-tools/agent-skills/best-practices#evaluation-and-iteration)

## Rule Description

List all required packages, libraries, and dependencies in the skill instructions. Verify they are available before use.

## Requirements

For skills that use external packages:
1. List all required packages explicitly
2. Include installation instructions if needed
3. Specify version requirements if applicable
4. Verify availability before use in scripts

## Rationale

**Claude cannot install packages**. If your skill requires a package that isn't available:

1. **Silent failures**: Scripts fail with cryptic import errors
2. **Wasted tokens**: Claude tries to debug unavailable dependencies
3. **User frustration**: "Why doesn't this work?"

**By documenting requirements**:
- Users can install packages before using the skill
- Claude knows what's expected to be available
- Clear error messages when dependencies are missing
- Better troubleshooting path

## Good Examples

**Python packages with installation**:

```markdown
## Required Packages

This skill requires the following Python packages:

```bash
pip install pdfplumber==0.10.0
pip install pillow>=9.0.0
```

**Check availability**:

```python
try:
    import pdfplumber
except ImportError:
    print("ERROR: pdfplumber not installed. Run: pip install pdfplumber")
    exit(1)
```
```

**Node.js packages**:

```markdown
## Dependencies

Install required packages:

```bash
npm install @google-cloud/bigquery
npm install googleapis
```

Required versions:
- @google-cloud/bigquery: ^7.0.0
- googleapis: ^118.0.0
```

**System tools**:

```markdown
## Required Tools

This skill requires:
- `jq` for JSON processing (install: `brew install jq` or `apt-get install jq`)
- `imagemagick` for image conversion (install: `brew install imagemagick`)
- `pandoc` for document conversion (https://pandoc.org/installing.html)

Verify installation:
```bash
command -v jq || echo "jq not found"
command -v convert || echo "imagemagick not found"
command -v pandoc || echo "pandoc not found"
```
```

## Bad Examples

**No package documentation**:

```python
import pdfplumber  # Where does this come from?
import some_obscure_library  # What is this?

# User has no idea what to install
```

**Vague requirements**:

```markdown
## Requirements

You'll need some Python packages for PDF processing.
```

**Undiscoverable dependencies**:

```markdown
# Hidden in script - user won't know until it fails
import rarely_used_package
```

## Detection

This rule checks for:

**Import statements without documentation**:
- Python: `import`, `from X import`
- Node.js: `require()`, `import ... from`
- Other: Language-specific imports

**Looks for package documentation**:
- Section headers: "Dependencies", "Requirements", "Installation", "Setup"
- Package lists with installation commands
- Version specifications

**Flags when**:
- Imports found in code but no dependency documentation
- Generic language like "install required packages" without specifics

## Severity

- **Default**: Warning
- **Strict mode**: Error

## What to Document

**Always document**:
- Third-party packages (not in standard library)
- System tools and command-line utilities
- External services or APIs requiring credentials
- Minimum version requirements

**Optional to document**:
- Standard library imports (e.g., Python's `os`, `json`)
- Built-in tools (e.g., `bash`, `grep` on Unix systems)
- Ubiquitous packages (e.g., `requests` in Python contexts)

## Verification Pattern

Include verification code to check dependencies:

**Python**:
```python
def check_dependencies():
    required = {
        'pdfplumber': 'pip install pdfplumber',
        'pandas': 'pip install pandas'
    }

    missing = []
    for package, install_cmd in required.items():
        try:
            __import__(package)
        except ImportError:
            missing.append(f"{package}: {install_cmd}")

    if missing:
        print("Missing dependencies:")
        for dep in missing:
            print(f"  - {dep}")
        exit(1)
```

**Bash**:
```bash
check_tools() {
    for tool in jq convert pandoc; do
        if ! command -v $tool &> /dev/null; then
            echo "ERROR: $tool not found"
            exit 1
        fi
    done
}
```

## Related Rules

- AS017: Scripts have error handling (dependency checks are error handling)
- AS009: MCP tool references (document which MCP servers are required)
