# AS014: Description Includes Usage Triggers

## Source

**Official Documentation**: [Writing Effective Descriptions - Agent Skills Best Practices](https://platform.claude.com/docs/en/agents-and-tools/agent-skills/best-practices#writing-effective-descriptions)

## Rule Description

The description field should include both what the skill does AND when to use it. Provide specific triggers and contexts that indicate when this skill is relevant.

## Requirements

Description should contain:
1. **What**: What the skill does
2. **When**: Triggers for when to use it

**Trigger phrases**:
- "Use when..."
- "Use for..."
- "Use if..."
- Similar contextual indicators

## Rationale

**The description enables skill discovery**. Claude uses it to choose from potentially 100+ available skills. Without usage triggers:

1. **Missed opportunities**: Claude doesn't know when to use the skill
2. **Wrong skill selected**: Similar skills compete, unclear which applies
3. **Manual activation**: User must explicitly invoke the skill

**With clear triggers**:
- Automatic skill discovery
- Correct skill selection
- Seamless integration into workflows

**Example**: A skill with description "Extract text from PDF files" might be overlooked when the user says "analyze this document" (if the document is a PDF). But "Extract text from PDF files. Use when working with PDFs or when the user mentions PDF, document extraction, or .pdf files" will trigger correctly.

## Good Examples

**PDF Processing**:

```yaml
description: Extract text and tables from PDF files, fill forms, merge documents. Use when working with PDF files or when the user mentions PDFs, forms, or document extraction.
```

What: Extract text/tables, fill forms, merge documents
When: PDF files, mentions of PDFs/forms/document extraction

**Excel Analysis**:

```yaml
description: Analyze Excel spreadsheets, create pivot tables, generate charts. Use when analyzing Excel files, spreadsheets, tabular data, or .xlsx files.
```

What: Analyze spreadsheets, pivot tables, charts
When: Excel files, spreadsheets, tabular data, .xlsx

**Git Commit Helper**:

```yaml
description: Generate descriptive commit messages by analyzing git diffs. Use when the user asks for help writing commit messages or reviewing staged changes.
```

What: Generate commit messages from diffs
When: Commit message help, staged changes review

**BigQuery Queries**:

```yaml
description: Query BigQuery datasets for sales and finance metrics. Use when analyzing company data, when the user mentions BigQuery, or when asking about sales/revenue/financial metrics.
```

What: Query BigQuery for sales/finance
When: Company data analysis, BigQuery mentions, sales/revenue/finance questions

## Bad Examples

**Missing triggers**:

```yaml
description: Processes PDF files
```

No "when" - Claude doesn't know when this applies.

**Too vague**:

```yaml
description: Helps with documents. Use when working with documents.
```

"Documents" is too broad - when is this NOT relevant?

**Only what, no when**:

```yaml
description: Extract text from images using OCR technology
```

What is clear, but when is missing. Add: "Use when working with scanned documents, images containing text, or when the user needs OCR."

**Generic triggers**:

```yaml
description: Analyzes data. Use when you need data analysis.
```

Too circular and generic.

## Detection

This rule checks the description field for:

**Presence of trigger phrases**:
- "use when"
- "use for"
- "use if"
- "use with"
- "when working with"
- "when the user mentions"
- "when asking about"

**Flags descriptions lacking**:
- No trigger phrases found
- Description under 50 characters (likely too brief)
- Generic terms without specific contexts

## Severity

- **Default**: Warning
- **Strict mode**: Error

## Crafting Effective Triggers

**Be specific**:
- ✓ "when the user mentions PDFs, .pdf files, or document extraction"
- ✗ "when working with files"

**Include synonyms**:
- ✓ "Excel, spreadsheets, .xlsx files, or tabular data"
- ✗ "Excel files"

**Consider user language**:
- ✓ "commit messages, staged changes, or git status"
- ✗ "git operations"

**Think about context**:
- ✓ "sales metrics, revenue data, or financial analysis"
- ✗ "data analysis"

## Trigger Categories

**File types**:
- "when working with PDF files, .pdf extensions"
- "when analyzing .xlsx, .csv, or Excel files"

**Keywords**:
- "when the user mentions 'commit', 'git', or 'staging'"
- "when asking about 'revenue', 'sales', or 'metrics'"

**Tasks**:
- "when generating reports"
- "when extracting data"
- "when validating forms"

**Tools/platforms**:
- "when using BigQuery"
- "when working with GitHub"
- "when accessing Slack"

## Related Rules

- AS002: Description validation (ensures description exists and is well-formed)
- AS003: Third-person voice (applies to description content)
