# AS010: Avoid Time-Sensitive Information

## Source

**Official Documentation**: [Avoid Time-Sensitive Information - Agent Skills Best Practices](https://platform.claude.com/docs/en/agents-and-tools/agent-skills/best-practices#avoid-time-sensitive-information)

## Rule Description

Don't include information that will become outdated, such as absolute dates or time-based conditionals.

## Requirements

Avoid references to specific calendar dates or time periods that will become obsolete:
- ✗ **Avoid**: "before August 2025", "after 2024", "in Q1 2025"
- ✓ **Good**: Use "old patterns" section for deprecated content

## Rationale

**Skills are long-lived**: A skill written in 2024 might still be used in 2027. Time-sensitive information becomes:
1. **Incorrect**: "Use method X before August 2025" → What if it's September 2025?
2. **Confusing**: "The new API launches in Q1 2025" → Is Q1 2025 "new" in 2026?
3. **Unmaintained appearance**: Stale dates make skills look abandoned

**Maintenance burden**: Every time-sensitive reference requires:
- Monitoring for when it becomes outdated
- Updates to remove or revise
- Risk of being forgotten and misleading users

## Good Examples

**Use "old patterns" section**:
```markdown
## Current approach

Use the v2 API: `api.example.com/v2/messages`

## Old patterns

<details>
<summary>Legacy v1 API (deprecated 2025-08)</summary>

The v1 API used: `api.example.com/v1/messages`

This API is no longer supported. Migrate to v2.
</details>
```

**Feature gates instead of dates**:
```markdown
# Conditional on feature availability, not date
if feature_available("new_api"):
    use_v2_api()
else:
    use_v1_api()
```

**Version-based guidance**:
```markdown
## Compatibility

- **Library v3.x and later**: Use `new_method()`
- **Library v2.x**: Use `old_method()` (deprecated)
```

## Bad Examples

**Absolute dates**:
```markdown
# BAD - Will be wrong after August 2025
If you're doing this before August 2025, use the old API.
After August 2025, use the new API.
```

**Relative time from publication**:
```markdown
# BAD - "Recently" loses meaning
The v2 API was recently released (January 2025).
Use this for all new projects.
```

**Quarter/year references**:
```markdown
# BAD - Q1 2025 won't be "upcoming" forever
The new authentication system launches in Q1 2025.
```

## Detection

This rule scans skill content for time-sensitive patterns:

**Absolute date patterns**:
- Month + year: "January 2025", "Aug 2024", "Q1 2025"
- Year-only: "in 2024", "after 2025", "before 2026"
- Full dates: "2025-08-15", "December 1st, 2024"

**Temporal qualifiers**:
- "before [date]", "after [date]"
- "as of [date]", "since [date]"
- "current year" (year-specific)

**Exceptions** (allowed):
- Inside `<details>` tags with "deprecated" or "legacy" (old patterns section)
- Copyright notices: "Copyright 2024"
- Version history: "Released 2024-01" in changelogs

## Severity

- **Default**: Warning
- **Strict mode**: Error

## Remediation

**Option 1: Remove date entirely**
```diff
- As of January 2025, use the new API
+ Use the new API (recommended)
```

**Option 2: Move to "old patterns"**
```diff
- Before June 2025: use method A
- After June 2025: use method B

+ Use method B (recommended)
+
+ ## Old patterns
+ <details>
+ <summary>Legacy method A (deprecated 2025-06)</summary>
+ Method A was replaced by method B in June 2025.
+ </details>
```

**Option 3: Use version gates**
```diff
- After March 2025, the new auth system is available
+ The new auth system is available in v3.0+
```

## Related Rules

- AS012: Use consistent terminology (helps avoid temporal language)
