# AS019: Workflows Use Numbered Steps and Checkboxes

## Source

**Official Documentation**: [Use Workflows for Complex Tasks - Agent Skills Best Practices](https://platform.claude.com/docs/en/agents-and-tools/agent-skills/best-practices#workflows-and-feedback-loops)

## Rule Description

Complex operations should be broken into clear, sequential steps using numbered lists or checkboxes. For particularly complex workflows, provide a checklist that Claude can copy and track.

## Requirements

For complex multi-step processes:
1. Break into sequential numbered steps
2. Use checkboxes (`- [ ]`) for trackable workflows
3. Make each step discrete and clear
4. Order steps logically

## Rationale

**Workflows prevent skipped steps and ensure completeness**. Without structured workflows:

1. **Steps missed**: Easy to skip important operations
2. **Order errors**: Steps performed out of sequence
3. **No tracking**: Hard to know what's complete
4. **Cognitive load**: Must remember entire process

**With numbered workflows**:
- Clear progress tracking
- Logical step ordering
- Nothing forgotten
- Easy to resume if interrupted

**Checkboxes enable active tracking**: Claude can copy the checklist and mark items complete, making progress visible.

## Good Examples

**Research workflow with checkboxes**:

````markdown
## Research Synthesis Workflow

Copy this checklist and track your progress:

```
Research Progress:
- [ ] Step 1: Read all source documents
- [ ] Step 2: Identify key themes
- [ ] Step 3: Extract supporting quotes
- [ ] Step 4: Organize by theme
- [ ] Step 5: Write synthesis
- [ ] Step 6: Add citations
- [ ] Step 7: Review for completeness
```
````

**Deployment workflow with numbers**:

```markdown
## Deployment Process

Follow these steps in order:

1. Run tests and verify all pass
2. Update version number in package.json
3. Build production bundle
4. Run security audit
5. Create git tag
6. Push to production
7. Verify deployment health
8. Update documentation
```

**Conditional workflow**:

```markdown
## Document Modification Workflow

1. Determine the modification type:

   **Creating new content?** → Follow "Creation workflow" below
   **Editing existing content?** → Follow "Editing workflow" below

2. **Creation workflow**:
   - [ ] Initialize new document structure
   - [ ] Add required metadata
   - [ ] Generate content
   - [ ] Validate against schema
   - [ ] Export to final format

3. **Editing workflow**:
   - [ ] Load existing document
   - [ ] Identify modification points
   - [ ] Apply changes
   - [ ] Validate after each change
   - [ ] Save updated document
```

**Data processing pipeline**:

```markdown
## Data Processing Pipeline

Complete these steps sequentially:

1. **Extract**: Load data from source
   - [ ] Connect to database
   - [ ] Execute query
   - [ ] Validate row count

2. **Transform**: Clean and reshape data
   - [ ] Remove null values
   - [ ] Normalize formats
   - [ ] Apply business rules

3. **Load**: Write to destination
   - [ ] Validate schema compatibility
   - [ ] Insert records
   - [ ] Verify record count matches

4. **Verify**: Check results
   - [ ] Run validation queries
   - [ ] Compare against expected output
   - [ ] Generate summary report
```

## Bad Examples

**No structure**:

```markdown
## Process

Do data extraction, cleaning, transformation, validation, and loading.
Make sure everything works correctly.
```

**Vague steps**:

```markdown
1. Prepare the data
2. Process it
3. Save the results
```

**Missing order**:

```markdown
## Tasks

- Clean data
- Save results
- Load from source (should be first!)
- Validate
- Transform
```

**Bullet points instead of sequence**:

```markdown
## Things to do

- Extract data
- Transform data
- Load data
- Validate results

(Which order? No checkboxes for tracking)
```

## Detection

This rule scans for workflow-like content and checks structure:

**Looks for workflow indicators**:
- Sections titled: "workflow", "process", "steps", "procedure"
- Multi-step processes
- Sequential operations

**Checks for proper structure**:
- Numbered lists (1. 2. 3.)
- Checkboxes (- [ ])
- Clear step descriptions

**Flags when**:
- Workflow content exists but no numbered lists or checkboxes
- Steps are unordered
- Workflow is present but structure is unclear

## Severity

- **Default**: Warning
- **Strict mode**: Error

## When to Use Workflows

**Definitely use workflows for**:
- Multi-step deployment processes
- Data processing pipelines
- Research and analysis procedures
- Document generation with multiple stages
- Complex configuration tasks
- Quality assurance checklists

**Optional for**:
- Simple 2-3 step processes
- Single-operation tasks
- Processes Claude handles natively

## Workflow Design Best Practices

**Make steps atomic**:
```markdown
✓ 1. Run unit tests
✓ 2. Run integration tests

✗ 1. Run all tests (too vague)
```

**Include validation**:
```markdown
1. Deploy to production
2. Verify health check endpoint returns 200
3. Check error logs for issues
```

**Add decision points**:
```markdown
3. Run tests
   - If tests fail → Fix issues and return to step 2
   - If tests pass → Continue to step 4
```

**Group related steps**:
```markdown
## Phase 1: Preparation
1. Load configuration
2. Validate environment
3. Check dependencies

## Phase 2: Execution
4. Process data
5. Generate output
6. Save results
```

## Related Rules

- AS011: Templates for output (workflows often include output templates)
