# AS012: Consistent Terminology

## Source

**Official Documentation**: [Use Consistent Terminology - Agent Skills Best Practices](https://platform.claude.com/docs/en/agents-and-tools/agent-skills/best-practices#use-consistent-terminology)

## Rule Description

Choose one term for each concept and use it consistently throughout the skill. Avoid mixing synonyms.

## Requirements

For key concepts in your skill, use the same term every time:
- Choose one term per concept
- Use it consistently in all contexts
- Avoid synonyms that mean the same thing

## Rationale

**Consistency helps Claude understand and follow instructions**. When you mix synonyms:

1. **Confusion**: Claude may think they're different concepts
2. **Inconsistent behavior**: Different terms trigger different mental models
3. **Reduced effectiveness**: Mixed terminology dilutes pattern recognition

**Example**: If you alternate between "API endpoint", "URL", "API route", and "path" to mean the same thing, Claude might:
- Treat them as distinct concepts
- Look for differences that don't exist
- Miss patterns that should be obvious

**With consistent terminology**:
- Clear mental model
- Reliable pattern matching
- Predictable behavior

## Good Examples

**Consistent - Single term per concept**:

```markdown
## API Endpoints

All API endpoints follow REST conventions.

To call an endpoint:
1. Build the endpoint URL
2. Send the request to the endpoint
3. Parse the endpoint response

Common endpoints:
- `/users` endpoint - user management
- `/posts` endpoint - content management
```

Always "endpoint", never mixing with "route", "URL", "path", etc.

**Consistent - Technical terms**:

```markdown
## Field Validation

Extract the field value from the form.
Check if the field is required.
Validate the field format.
Save the field to the database.
```

Always "field", never mixing with "element", "control", "box", "input".

## Bad Examples

**Inconsistent - Multiple synonyms**:

```markdown
## API Access

Call the API endpoint to get data.
Send a request to the URL.
The API route returns JSON.
Hit the path to retrieve results.
```

Four different terms (endpoint, URL, route, path) for the same concept.

**Inconsistent - Operation verbs**:

```markdown
## Data Processing

Extract text from the PDF.
Pull the metadata from the file.
Get the page count.
Retrieve the author information.
Grab the creation date.
```

Five different verbs (extract, pull, get, retrieve, grab) for the same operation.

## Detection

This rule scans for common synonym groups and flags when multiple synonyms appear in the same skill:

**Common synonym groups checked**:
- API access: endpoint, URL, route, path, URI
- Data extraction: extract, pull, get, retrieve, fetch, grab, obtain
- Form elements: field, input, control, element, box
- Documentation: guide, documentation, reference, manual, docs
- Configuration: setting, option, configuration, parameter, preference
- Storage: save, store, persist, write, record

## Severity

- **Default**: Warning
- **Strict mode**: Error

## Remediation

1. **Identify your terms**: List all key concepts in your skill
2. **Choose primary terms**: Pick one term for each concept
3. **Find and replace**: Search for synonyms and replace with primary term
4. **Document choices**: Optionally note terminology choices in README

**Example remediation**:

```diff
- Extract text from the PDF
- Pull the metadata
- Get the page count
+ Extract text from the PDF
+ Extract the metadata
+ Extract the page count
```

## Exceptions

**Acceptable variation**:
- Technical accuracy (HTTP "method" vs "verb" - both correct in context)
- Avoiding repetition in prose ("The user... they..." not "The user... the user...")
- Domain-specific terminology (database "row" vs API "record" - different contexts)

## Related Rules

- AS003: Use third-person voice (both ensure consistent language)
