# AS011: Templates for Output-Generating Skills

## Source

**Official Documentation**: [Template Pattern - Agent Skills Best Practices](https://platform.claude.com/docs/en/agents-and-tools/agent-skills/best-practices#template-pattern)

## Rule Description

Skills that generate structured output (reports, analyses, documents, API responses) should provide templates showing the expected format.

## Requirements

For skills that produce formatted output, include at least one template or example showing:
- Expected structure
- Required sections
- Format specifications

## Rationale

**Templates guide Claude's output structure**. When generating content like reports, analyses, or API responses, Claude needs to know the expected format. Without templates:

1. **Output varies**: Each invocation produces different structure
2. **Users must reformat**: Generated content requires manual restructuring
3. **Quality inconsistent**: Critical sections may be missing

**With templates**:
- Consistent output every time
- All required sections present
- Proper formatting from the start

**Strictness levels**: Match template strictness to requirements:
- **Strict** (API responses, data formats): "ALWAYS use this exact template"
- **Flexible** (reports, analyses): "Use this as a starting point, adapt as needed"

## Good Examples

**Strict template for API responses**:

````markdown
## Report Structure

ALWAYS use this exact template structure:

```markdown
# [Analysis Title]

## Executive Summary
[One-paragraph overview of key findings]

## Key Findings
- Finding 1 with supporting data
- Finding 2 with supporting data
- Finding 3 with supporting data

## Recommendations
1. Specific actionable recommendation
2. Specific actionable recommendation
```
````

**Flexible template for analysis**:

````markdown
## Analysis Format

Use this structure as a starting point:

```markdown
# Analysis Title

## Overview
[Context and scope]

## Data Analysis
[Your analysis here]

## Conclusions
[Key takeaways]
```

Adapt sections as needed for your specific analysis.
````

**Code output template**:

````markdown
## Generated Report Format

```python
def generate_report(data):
    return {
        "summary": "...",
        "metrics": {
            "total": 0,
            "average": 0
        },
        "details": [...]
    }
```
````

## Bad Examples

**No template provided**:

```markdown
## Report Generation

Generate a comprehensive report analyzing the data.
Include key findings and recommendations.
```

**Vague guidance**:

```markdown
## Output Format

Create a well-formatted report with all necessary sections.
```

## Detection

This rule checks if skills that appear to generate output include templates or examples:

**Heuristics for output-generating skills**:
- Description mentions: "generate", "create", "produce", "format", "report", "document"
- Content mentions: "output", "result", "response", "format"

**Looks for templates**:
- Code blocks with structure
- Example output sections
- Template sections with "structure", "format", "template"

## Severity

- **Default**: Warning
- **Strict mode**: Error

## When Templates Are Required

**Definitely need templates**:
- Report generation skills
- Document creation skills
- API response formatting skills
- Data export skills
- Analysis output skills

**Templates optional**:
- Data processing skills (input → transformation → return)
- Query/search skills (user sees results directly)
- Tool execution skills (output handled by tool)

## Related Rules

- AS019: Workflows use numbered steps (templates often part of workflows)
