# AS016: Avoid Reserved Words in Names

## Source

**Official Documentation**: [Naming Conventions - Agent Skills Best Practices](https://platform.claude.com/docs/en/agents-and-tools/agent-skills/best-practices#naming-conventions)

## Rule Description

Skill names cannot contain reserved words: "anthropic" or "claude". This applies to the name field and is also enforced by AS001.

## Requirements

Do not use these words in skill names:
- "anthropic" (case-insensitive)
- "claude" (case-insensitive)

## Rationale

**Reserved words protect branding and prevent confusion**:

1. **Official branding**: "Anthropic" and "Claude" are reserved for official products
2. **Source confusion**: Users might think these are official Anthropic skills
3. **Trust issues**: Unofficial use of brand names erodes trust
4. **Namespace protection**: Reserves space for future official skills

**Example issues**:
- `claude-helper` - Implies official Anthropic support
- `anthropic-tools` - Suggests official tooling
- `super-claude` - Misrepresents relationship with official product

## Good Examples

**Descriptive without reserved words**:

```yaml
name: processing-pdfs
```

```yaml
name: analyzing-spreadsheets
```

```yaml
name: database-queries
```

```yaml
name: code-review-helper
```

**Even when tempted to use brand**:

```yaml
# NOT: claude-assistant
name: ai-assistant-tools
```

```yaml
# NOT: anthropic-api-helper
name: api-integration-tools
```

## Bad Examples

**Contains "claude"**:

```yaml
name: claude-helper
```

```yaml
name: super-claude
```

```yaml
name: claude-tools
```

```yaml
name: my-claude-skill
```

**Contains "anthropic"**:

```yaml
name: anthropic-helper
```

```yaml
name: anthropic-tools
```

```yaml
name: anthropic-integration
```

## Detection

This rule checks the `name` field for reserved words:

**Case-insensitive search** for:
- "anthropic" (matches: Anthropic, ANTHROPIC, anthropic)
- "claude" (matches: Claude, CLAUDE, claude)

**In any position**:
- Beginning: `claude-helper`
- Middle: `super-claude-tools`
- End: `helper-claude`

## Severity

- **Default**: Warning
- **Strict mode**: Error

## Why This Matters

**Brand protection**:
- Anthropic owns the trademarks
- Unauthorized use creates legal issues
- Official products need protected namespace

**User trust**:
- Clear distinction between official and community skills
- No confusion about support and maintenance
- Transparent about skill origins

**Future compatibility**:
- Anthropic may release official skills
- Reserved namespace prevents conflicts
- Clear upgrade path for official versions

## Alternative Naming

Instead of using reserved words, describe functionality:

```diff
- claude-helper
+ ai-assistant-helper

- anthropic-api-tools
+ api-integration-tools

- super-claude
+ enhanced-assistant

- claude-docs
+ documentation-generator
```

## Related Rules

- AS001: Name format validation (also checks reserved words)
- AS015: Prefer gerund naming (suggests alternative patterns)
