# AS020: Table of Contents Completeness

## Source

**Official Documentation**: [Structure Longer Reference Files - Agent Skills Best Practices](https://platform.claude.com/docs/en/agents-and-tools/agent-skills/best-practices#structure-longer-reference-files-with-table-of-contents)

## Rule Description

For files with a table of contents, ensure the TOC matches the actual headers in the document. All major sections should be listed.

## Requirements

When a TOC is present:
1. All level 2 headers (##) should be in the TOC
2. TOC links should point to existing headers
3. Header names in TOC should match actual headers
4. TOC should be near the beginning of the file

## Rationale

**An incomplete TOC is worse than no TOC**. When the TOC doesn't match content:

1. **Broken navigation**: Links point to non-existent sections
2. **Missing sections**: Important content not discoverable
3. **Confusion**: TOC promises content that isn't there (or vice versa)
4. **Wasted effort**: Claude navigates to wrong sections

**A complete, accurate TOC**:
- Enables quick navigation
- Shows full scope of content
- Links work correctly
- Builds trust

## Good Examples

**Complete TOC matching content**:

````markdown
# API Reference

## Table of Contents

1. [Authentication](#authentication)
2. [Core Methods](#core-methods)
   - [Create Document](#create-document)
   - [Update Document](#update-document)
   - [Delete Document](#delete-document)
3. [Advanced Features](#advanced-features)
4. [Error Handling](#error-handling)
5. [Rate Limits](#rate-limits)

## Authentication

[Content about authentication]

## Core Methods

### Create Document

[Create documentation]

### Update Document

[Update documentation]

### Delete Document

[Delete documentation]

## Advanced Features

[Advanced content]

## Error Handling

[Error handling content]

## Rate Limits

[Rate limit content]
````

All TOC entries have corresponding sections.

**Comprehensive TOC**:

```markdown
# User Guide

## Contents

- [Overview](#overview)
- [Installation](#installation)
- [Configuration](#configuration)
- [Usage](#usage)
- [Troubleshooting](#troubleshooting)
- [FAQ](#faq)

[All sections present below]
```

## Bad Examples

**Missing sections**:

```markdown
## Table of Contents

1. [Authentication](#authentication)
2. [Core Methods](#core-methods)

## Authentication
...

## Core Methods
...

## Advanced Features  ← NOT in TOC!
...

## Error Handling      ← NOT in TOC!
...
```

**Broken links**:

```markdown
## Contents

- [Installation](#installing)  ← Link mismatch!
- [Configuration](#config)     ← Link mismatch!

## Installation  ← Actual header
...

## Configuration  ← Actual header
...
```

**Mismatched names**:

```markdown
## Table of Contents

- [Auth](#authentication)
- [API Methods](#core-methods)

## Authentication  ← Name doesn't match "Auth"
...

## Core Methods  ← Name doesn't match "API Methods"
...
```

## Detection

This rule validates TOC completeness:

**For files with TOC**:
1. Extract TOC links (markdown anchors: `[text](#anchor)`)
2. Extract actual headers from content
3. Compare: Are all headers in TOC?
4. Check: Do TOC links point to existing headers?

**Flags when**:
- TOC exists but major headers are missing from it
- TOC links don't match actual header anchors
- TOC has entries without corresponding headers

## Severity

- **Default**: Warning
- **Strict mode**: Error

## TOC Best Practices

**Include all major sections**:
```markdown
## Table of Contents

✓ All level 2 headers (##)
✓ Important level 3 headers (###) if document is complex
✗ Don't include every tiny subsection (too cluttered)
```

**Match header text exactly**:
```markdown
✓ [Error Handling](#error-handling)
## Error Handling

✗ [Errors](#error-handling)
## Error Handling  (mismatch)
```

**Use proper anchor format**:
```markdown
✓ [Rate Limits](#rate-limits)        → ## Rate Limits
✓ [API Methods](#api-methods)        → ## API Methods
✗ [Rate Limits](#RateLimits)         → Wrong anchor format
```

**Keep TOC updated**:
```markdown
When you add a section:
1. Add the header to the document
2. Add the entry to the TOC
3. Verify the link works
```

## Anchor Format

**Standard GitHub/CommonMark format**:
- Lowercase all letters
- Replace spaces with hyphens
- Remove special characters (except hyphens)

**Examples**:
```markdown
## Authentication and Setup
→ #authentication-and-setup

## Core Methods (CRUD)
→ #core-methods-crud

## API v2.0
→ #api-v20
```

## Relationship with AS008

**AS008**: Requires TOC for files > 100 lines
**AS020**: Validates that existing TOCs are complete

Together they ensure:
- Long files have TOCs (AS008)
- TOCs are accurate and complete (AS020)

## Related Rules

- AS008: TOC required for files > 100 lines (ensures TOC exists)
- AS006: References one level deep (TOC helps navigate reference files)
